### DISCUSSION TITLE
[Share] Open PDFs with the system default viewer

### DISCUSSION THREAD

**Discussion by YDX-2147483647** [windingwind/zotero-actions-tags#359] [Action Scripts]
### Description

Zotero provides a setting to open PDFs with either its _internal_ viewer or the _system_ default. You can set it to _internal_, and use this script to open PDFs with the system default.

### Event

None

### Operation

Script

### Data

```javascript
/**
 * Open PDFs with the system default
 * @author Y.D.X.
 * @usage Select an item in the Library, then trigger this script.
 * @link https://github.com/windingwind/zotero-actions-tags/discussions/359
 * @see https://github.com/windingwind/zotero-actions-tags/discussions/359
 */

// Modified from zotero-open-pdf.
// https://github.com/retorquere/zotero-open-pdf/blob/b7cd70d04c4dc0791a5f49542552a5231062637d/lib.ts#L198-L204

if (!items?.length) return;

items.forEach(async (item) => {
  const attachment = item.isAttachment()
    ? item
    : await item.getBestAttachment();
  if (attachment?.attachmentPath.toLowerCase().endsWith(".pdf")) {
    Zotero.launchFile(attachment.getFilePath());
  }
});
```

### Anything else

Alternative solutions:

- https://github.com/retorquere/zotero-open-pdf
  A plugin that modifies the context menu.
- #132
  A similar script, but you have to explicitly specify a PDF viewer.
- https://github.com/wshanks/Zutilo
  A plugin for keyboard shortcuts. Zotero 7 is not supported yet.

**Comment by noppelmax** [comment]
Is there a way to define the system pdf viewer hard coded? I am currently using two. [Evince](https://wiki.ubuntuusers.de/Evince/) and [Okular](https://okular.kde.org/de/) for different purposes and I would like to have two actions "Open in evince" and "Open in okular".

So instead of `Zotero.launchFile` I would need something like `System.call("evince ".attachment.getFilePath()).detach()`.

**Reply by mjthoraval** [reply]
Should be this one: https://github.com/windingwind/zotero-actions-tags/discussions/132

**Reply by noppelmax** [reply]
Find my solution here:
```javascript
if (!items?.length) return;

if (!item && !collection) return "[Open in Okular] item is empty";
if (collection) return "[Open in Okular] Cannot open a collection";

async function getAttachmentPath(item) {
  if (item.isAttachment() && !item.isNote()) {
    return await item.getFilePathAsync();
  } else if (item.isRegularItem() && !item.isAttachment()) {
    let attachment = await item.getBestAttachment();
    return await attachment.getFilePathAsync();
  }
  return null;
}

items.forEach(async (item) => {
  const filePath = await getAttachmentPath(item);
  if (filePath.toLowerCase().endsWith(".pdf")) {
    let applicationPath = "/usr/bin/okular";
    let args = [filePath];
    Zotero.Utilities.Internal.exec(applicationPath, args);
  }
});
```
